browser/ui/webui/brave_sanitized_image_source.cc

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/ui/webui/brave_sanitized_image_source.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/memory/ref_counted_memory.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_split.h"
+#include "base/strings/string_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/task/task_traits.h"
+#include "base/task/thread_pool.h"
+#include "brave/brave_domains/service_domains.h"
+#include "brave/components/brave_private_cdn/private_cdn_helper.h"
+#include "chrome/browser/profiles/profile.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/storage_partition.h"
+#include "net/base/url_util.h"
+#include "net/traffic_annotation/network_traffic_annotation.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "third_party/skia/include/core/SkBitmap.h"
+#include "ui/gfx/codec/png_codec.h"
+#include "ui/gfx/geometry/size.h"
+#include "ui/gfx/image/image.h"
+#include "ui/gfx/image/image_skia_operations.h"
+#include "url/url_util.h"
+
+namespace {
+constexpr char kUrlKey[] = "url";
+constexpr char kTargetSizeKey[] = "target_size";
+constexpr char kChromeUIBraveImageURL[] = "chrome://brave-image/";
+constexpr char kChromeUIBraveImageHost[] = "brave-image";
+
+std::map<std::string, std::string> ParseParams(std::string_view param_string) {
+  url::Component query(0, param_string.size());
+  url::Component key;
+  url::Component value;
+  constexpr int kMaxUriDecodeLen = 2048;
+  std::map<std::string, std::string> params;
+  while (url::ExtractQueryKeyValue(param_string, &query, &key, &value)) {
+    url::RawCanonOutputW<kMaxUriDecodeLen> output;
+    url::DecodeURLEscapeSequences(param_string.substr(value.begin, value.len),
+                                  url::DecodeURLMode::kUTF8OrIsomorphic,
+                                  &output);
+    params.insert({std::string(param_string.substr(key.begin, key.len)),
+                   base::UTF16ToUTF8(output.view())});
+  }
+  return params;
+}
+
+scoped_refptr<base::RefCountedMemory> EncodeImage(
+    SkBitmap bitmap,
+    const gfx::Size& target_size) {
+  if (!target_size.IsEmpty() && bitmap.width() > 0 && bitmap.height() > 0) {
+    // Resize to the target size, keeping the aspect ratio:
+    auto target_width = target_size.width();
+    auto target_height = target_size.height();
+    auto width_with_ratio = target_height * bitmap.width() / bitmap.height();
+    if (width_with_ratio > target_width) {
+      target_width = width_with_ratio;
+    } else {
+      target_height = target_width * bitmap.height() / bitmap.width();
+    }
```

Comment:
This doesn't seem bad, but it doesn't quite work how I would've expected passing in a box, I'd expected the resulting image to fit in the box - instead we resize the height to fit.

Reply:
I guess this is probably better for making sure the image isn't pixellated though

Reply:
Well, we can add cropping or clarify the explanation.
The idea of this code is to get the same pixels as we have now in `master`: we resized to a the dimensions to be perfectly cropped to the target size, without extra up/downscaling.